packages/babel-helper-builder-react-jsx-experimental/src/index.js

```diff
@@ -0,0 +1,844 @@
+  // step while we deprecate key spread from props. Afterwards,
+  // we will stop using createElement in the transform.
+  function shouldUseCreateElement(path) {
+    const openingPath = path.get("openingElement");
+    const attributes = openingPath.node.attributes;
+
+    let seenPropsSpread = false;
+    for (let i = 0; i < attributes.length; i++) {
+      const attr = attributes[i];
+      if (
+        seenPropsSpread &&
+        t.isJSXAttribute(attr) &&
+        attr.name.name === "key"
+      ) {
+        return true;
+      } else if (t.isJSXSpreadAttribute(attr)) {
+        seenPropsSpread = true;
+      }
+    }
+    return false;
+  }
+
+  function createIdentifierName(path, name, importName) {
+    if (isModule(path) && importName) {
+      const identifierName = `${importName[name]}`;
+      return identifierName;
+    } else {
+      return `${importName}.${name}`;
+    }
+  }
+
+  function getImportNames(parentPath) {
+    const imports = {};
+
+    parentPath.traverse({
+      JSXElement(path) {
+        const openingPath = path.get("openingElement");
+        openingPath.parent.children = t.react.buildChildren(openingPath.parent);
+
+        if (shouldUseCreateElement(path)) {
+          imports.createElement = true;
+        } else if (path.node.children.length > 1) {
+          const importName = options.development ? "jsxDEV" : "jsxs";
+          imports[importName] = true;
+        } else {
+          const importName = options.development ? "jsxDEV" : "jsx";
+          imports[importName] = true;
+        }
+      },
+
+      JSXFragment(path) {
+        imports.Fragment = true;
+
+        const openingPath = path.get("openingElement");
+        openingPath.parent.children = t.react.buildChildren(openingPath.parent);
+
+        if (path.node.children.length > 1) {
+          const importName = options.development ? "jsxDEV" : "jsxs";
+          imports[importName] = true;
+        } else {
+          const importName = options.development ? "jsxDEV" : "jsx";
+          imports[importName] = true;
+        }
+      },
+    });
+    return imports;
+  }
+
+  function hasJSX(parentPath) {
+    let fileHasJSX = false;
+    parentPath.traverse({
+      JSXElement(path) {
+        fileHasJSX = true;
+        path.stop();
+      },
+
+      JSXFragment(path) {
+        fileHasJSX = true;
+        path.stop();
+      },
+    });
+
+    return fileHasJSX;
+  }
+
+  function addAutoImports(path, state) {
+    if (isModule(path)) {
+      // import {jsx} from "react";
+      // import {createElement} from "react";
+      const imports = getImportNames(path, state);
+      const importMap = {};
+
+      Object.keys(imports).forEach(importName => {
+        importMap[importName] = addNamed(path, importName, state.source).name;
+      });
+
+      return importMap;
+    } else {
+      // var _react = require("react")
+      return addNamespace(path, state.source, {
+        importedInterop: "uncompiled",
+      }).name;
+    }
+  }
```

Comment:
Currently, this function returns sometimes an object, sometimes a string. It would be better to unify them.

You can simplify this function like this:
```js
function addAutoImports(path, state) {
  const importMap = {};
  Object.keys(imoprts).forEach(importName => {
    importMap[importName] = addNamed(path, imoprtName, state.source, {
      importedInterop: "uncompiled",
      ensureLiveReference: true,
    });
  });
}
```

And the instead of `createIdentifierName(path, name, importName)` use `t.cloneNode(importName[name])`.

Reply:
This won't exactly work because it will add a separate `require` for every import.

Reply:
Oh ok.
I'd like to make `addNamed` work better with multiple named imports in this case, but I/you/anyone can do it in a separate PR.

Reply:
Okay sounds good. I'll just leave this for now then and update after `addNamed` is fixed.

Reply:
We'll probably merge this PR without this change, to avoid blocking it.